### PR TITLE
SDI-33 Added concept of batch context

### PR DESCRIPTION
**Pull request by andymarke**
The purpose of this PR to to add the concept of a context for migration run, so that any filters and a unique migration is passed around the messaging callbacks

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/listeners/MigrationMessageListener.kt [hunk 1]
```diff
@@ -22,16 +25,23 @@ class MigrationMessageListener(
...
-  data class MigrationMessage(
+  open class MigrationMessage<T>(
     val type: Messages,
-    val body: Any? = null
+    open val context: MigrationContext<T>
   )
 
   private inline fun <reified T> String.fromJson(): T = objectMapper.readValue(this, T::class.java)
 }
+
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/listeners/MigrationMessageListener.kt:41]: requires more work below (or a rethink)

File: helm_deploy/hmpps-prisoner-from-nomis-migration/values.yaml [hunk 2]
```diff
@@ -19,6 +19,7 @@ generic-service:
           return 401;
         }
     host: app-hostname.local    # override per environment
+    contextColour: green
     tlsSecretName: prisoner-from-nomis-cert
```
Comment (andymarke) [inline helm_deploy/hmpps-prisoner-from-nomis-migration/values.yaml:23]: unrelated, but moved to generic config

File: src/main/resources/banner.txt [hunk 3]
```diff
@@ -1,13 +1,14 @@
...
+██╔═══╝░██╔══██╗██║░╚═══██╗██║░░██║██║╚████║██╔══╝░░██╔══██╗  ██╔══╝░░██╔══██╗██║░░██║██║╚██╔╝██║
+██║░░░░░██║░░██║██║██████╔╝╚█████╔╝██║░╚███║███████╗██║░░██║  ██║░░░░░██║░░██║╚█████╔╝██║░╚═╝░██║
+╚═╝░░░░░╚═╝░░╚═╝╚═╝╚═════╝░░╚════╝░╚═╝░░╚══╝╚══════╝╚═╝░░╚═╝  ╚═╝░░░░░╚═╝░░╚═╝░╚════╝░╚═╝░░░░░╚═╝
+
+███╗░░██╗░█████╗░███╗░░░███╗██╗░██████╗  ███╗░░░███╗██╗░██████╗░██████╗░░█████╗░████████╗██╗░█████╗░███╗░░██╗
+████╗░██║██╔══██╗████╗░████║██║██╔════╝  ████╗░████║██║██╔════╝░██╔══██╗██╔══██╗╚══██╔══╝██║██╔══██╗████╗░██║
+██╔██╗██║██║░░██║██╔████╔██║██║╚█████╗░  ██╔████╔██║██║██║░░██╗░██████╔╝███████║░░░██║░░░██║██║░░██║██╔██╗██║
+██║╚████║██║░░██║██║╚██╔╝██║██║░╚═══██╗  ██║╚██╔╝██║██║██║░░╚██╗██╔══██╗██╔══██║░░░██║░░░██║██║░░██║██║╚████║
+██║░╚███║╚█████╔╝██║░╚═╝░██║██║██████╔╝  ██║░╚═╝░██║██║╚██████╔╝██║░░██║██║░░██║░░░██║░░░██║╚█████╔╝██║░╚███║
+╚═╝░░╚══╝░╚════╝░╚═╝░░░░░╚═╝╚═╝╚═════╝░  ╚═╝░░░░░╚═╝╚═╝░╚═════╝░╚═╝░░╚═╝╚═╝░░╚═╝░░░╚═╝░░░╚═╝░╚════╝░╚═╝░░╚══╝
```
Comment (steverendell) [inline src/main/resources/banner.txt:14]: I thought this font looked cool .. but found it just displays in ???s in my local console !

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/listeners/MigrationMessageListener.kt [hunk 4]
```diff
@@ -22,16 +25,20 @@ class MigrationMessageListener(
...
-  data class MigrationMessage(
+  class MigrationMessage<T>(
     val type: Messages,
-    val body: Any? = null
+    val context: MigrationContext<T>
   )
 
-  private inline fun <reified T> String.fromJson(): T = objectMapper.readValue(this, T::class.java)
+  private inline fun <reified T> String.fromJson(): T =
+    objectMapper.readValue(this, object : TypeReference<T>() {})
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/listeners/MigrationMessageListener.kt:40]: I could be wrong but I found that if you use a "Kotlin" objectMapper there is a clever single argument readValue() method.
i.e. use  objectMapper = ObjectMapper().registerModule(KotlinModule.Builder().build())

**Review by lbennett-moj — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/listeners/MigrationMessageListener.kt (hunk 4 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/listeners/MigrationMessageListener.kt:40]: cool, I'll looked at that

File: src/main/resources/banner.txt (hunk 3 above)
Comment (andymarke) [reply src/main/resources/banner.txt:14]: Weird that works for me

Comment (andymarke) [reply src/main/resources/banner.txt:14]: I wonder if it is a Windows thing

**Pull request merged**